### DISCUSSION TITLE
What are the equivalent keywords to C#'s `in`, `out` and `ref`?

### DISCUSSION THREAD

**Discussion by ghost** [pascalabcnet/pascalabcnet#3014] [Q&A] [answered]
I encountered them when reading the pinvoke document. Please help.

**Comment by SunSerega** [comment, marked as answer]
They are all types of references. In Pascal, you currently only have a kinda ugly and old syntax equivalent to `ref`:
```
procedure p1(var x: byte) := exit;
```
It's meant to look like you're passing in the variable instead of the value.
```
##
procedure p1(var x: byte) := x := 2;

var x: byte := 1;
Println(x);
p1(x);
Println(x);
```
Here `p1` changes the variable's value because `p1` took in that variable, not its values (as it usually would).
But this explanation breaks down when what you pass isn't a variable but, for instance, an element of an array.

And this kind of behavior can only be achieved by passing reference to a value instead of an actual value.
So in our age `var`-parameters are just said to be parameters-by-referece.

---

Now, `in` and `out` are modifiers for `ref`. `out` is [planned](https://github.com/pascalabcnet/pascalabcnet/issues/2828), but neither currently work.
But they are sugar (not necessarily syntax sugar), so them being supported on the compiler level is a convenience.

`out` means that the function will set at least something to that `ref` parameter before exiting.
C# uses this restriction to instead allow declaring variable right inside function call:
```
if (dictionary.TryGetValue(key, out var value))
    // value variable only exists inside this if
```
(and this ability is why #2828 exists)

`in` means that the function only takes the referenced value `in` and doesn't modify it.
But currently, you can't combine `const` and `var` before the parameter in Pascal, so you don't have a way to make the compiler check if your code is correct like that.

You can, however, superficially specify that your function's `var`-parameter is actually `out`- or `in`-parameter. Thought, it would be just a promise, which you can break.
In C#, `(out byte x)` turns into `([Out] ref byte x)`, and the same thing for the `in`-parameters.


```
## uses System.Runtime.InteropServices;

// Need &, because "in" is a keyword
procedure p1([Out] var a: byte; [&In] var b: byte);
begin
  // Compiler allows to read "a" before setting anything
  // And allows to set "b", even though
  b := a;
end;
```
This is useful:
1. For `external` subprograms, to specify marshaling direction, or
2. If your subprogram will be used from another .Net language, where `in` and `out` are fully supported

---

Another way to look at it: `var`-parameters are managed pointers.
```
##

procedure p1(var x: byte) := x := 1;
procedure p2(x: ^byte) := x^ := 1;

var x: byte := 1;
p1(x);
p2(@x);
```
`p1` and `p2` here are equivalent. Except it wouldn't always hold:
```
##

procedure p1(var x: byte) := x := 1;
procedure p2(x: ^byte) := x^ := 1;

var a := new byte[1];
p1(a[0]); // OK
p2(@a[0]); // Instant memory leak
```
`GC` can move class contents in the memory (to compact it after deleting small pieces of memory stored before it).
But checking far enough to know when the pointer will stop being used is hard. So instead the compiler just perma-locks the array in memory.
You can do it manually if you *really* want to, and then you'll be able to unlock it:
```
## uses System.Runtime.InteropServices;

procedure p1(var x: byte) := x := 1;
procedure p2(x: ^byte) := x^ := 1;

var a := new byte[1];

p1(a[0]);

var gc_hnd := GCHandle.Alloc(a, GCHandleType.Pinned);
try
  p2(gc_hnd.AddrOfPinnedObject.ToPointer);
finally
  gc_hnd.Free;
end;
```
But besides being bulky, this is also a bit slow.

---

I guess this is also relevant to unmanaged code. So let me show you some forbidden magic:
```
procedure p1(var x: byte); external '.dll';
procedure p1(a: array of byte);
begin
  if (a=nil) or (a.Length=0) then
    p1(PByte(nil)^) else
    p1(a[0]);
end;

procedure p1<T>(var x: T) := p1(PByte(pointer(@x))^);
procedure p1<T>(a: array of T); where T: record;
type PT = ^T;
begin
  if (a=nil) or (a.Length=0) then
    p1(PT(nil)^) else
    p1(a[0]);
end;
```
The `PByte(pointer(@x))^` is actually a no-op. It takes a reference of any type and passes it into another overload as if it were a reference to a byte.
No pointers are created in the process. Thought you need `x` to already be a `var`-parameter. If you try to do it to an array element - you'll again get a memory leak.

`PT(nil)^` creates a reference that points to nothing.
In C++ this would be a horrible UB. You'd basically need to create another `external` overload that takes in a pointer, just to pass `nil` to it.
But in .Net it just works - the default marshaler handles all kinds of references pretty well:
It locks class contents in memory. And does that much more efficiently than you can manually with `GCHandle`.
But if the reference doesn't point to managed memory (points to unmanaged one or is nil) - the default marshaler doesn't touch it.
